feat: hosts payed by actual time hosting a slot

AuHau opened the pull request:
This PR bundles few changes around payout and withdraws:

 - Hosts are paid by their actual time hosting the slot (also relevant for repair)
 - Client can withdraw remaining funds for finished requests
 - Client can withdraw all original funds for failed requests
 - When a slot is "forcible freed" and the slot would not be filled again, but the request would successfully finish - client can withdraw the funds for this slot
 
I am not so happy about the testing because it is pretty hard to correctly calculate the amounts that should be returned due to Hardhat's "auto" mining of blocks every time there is a transaction, which advances the clock and hence shifts the reward calculations. I made it work, but not great. My brain is fried though, so I will leave it for now and check back on it tomorrow. If you have any suggestions on how to track the funds better, I would love to hear them!

Closes #98
Closes #100 
Closes #102 
Closes #156

emizzle commented on test/Marketplace.test.js:
```diff
@@ -652,14 +697,41 @@ describe("Marketplace", function () {
+      await marketplace.withdrawFunds(slot.request)
+      const endBalance = await token.balanceOf(client.address)
+      expect(
+        maxPrice(request) -
+          (ACCOUNT_STARTING_BALANCE - endBalance) -
+          (request.expiry / 2) * request.ask.reward - // we have moved one slot to mid-expiry window
+          17 * request.ask.reward // There was 17 transaction for filling/advancing etc
```
If you took a `startBalance` sample before calling `withdrawFunds`, shouldn't `expectedHostPayout` be equal to `(endBalance - startBalance)`? The idea being that if all four slots were filled for the entire duration, the client should have no refunds, and `(endBalance - startBalance)` would be 0. However, since one slot is freed after starting, and we expect a refund, the difference should be equal to the refund.

emizzle replied:
Hmm actually I think it would not be 0 if there were no refunds as it depends when the slots were filled, and each transaction produces a block that is advanced by 1 second. So maybe instead of having a function `waitUntilStarted` with a parameter to not start certain slots, maybe there could be a function that fills certain slots and returns an array of start times? Those start times could be used to calculate how much each slot should be paid.

AuHau replied:
Yeah, this is something I more or less did. Please check the reworked test.

emizzle commented on test/Marketplace.test.js:
```diff
@@ -652,14 +697,41 @@ describe("Marketplace", function () {
+      await expect(marketplace.freeSlot(id)).to.emit(marketplace, "SlotFreed")
+      await waitUntilFinished(marketplace, requestId(request))
+
+      const expectedHostPayout =
+        ((await marketplace.requestEnd(requestId(request))).toNumber() -
+          filledAt) *
+        request.ask.reward
```
Should this use `payoutForDuration`?

emizzle commented on test/marketplace.js:
```diff
@@ -14,11 +14,18 @@ async function waitUntilCancelled(request) {
 
   for (let i = 0; i < request.ask.slots; i++) {
-    await contract.fillSlot(requestId(request), i, proof)
+    if (!ignoreSlotIndexes.includes(i))
```
How can the contract be started if not all slots are filled? Maybe this needs its own function?

emizzle replied:
Right, but the contract won't be started if the optional parameter is used to exclude a slot from being filled...

AuHau replied:
Well yes, but then you should know what you are doing if you are specifying that optional parameter...

markspanbroek replied:
🟠 I find this confusing as well. Why not add a `waitUntilSlotsFilled(indices)`? You can even use it to implement `waitUntilStarted()`

emizzle commented on test/Marketplace.test.js:
```diff
@@ -464,20 +464,30 @@ describe("Marketplace", function () {
-    it("pays the host when contract has finished and returns collateral", async function () {
-      await waitUntilStarted(marketplace, request, proof, token)
+    it("pays the host when contract has finished for time he hosted the request and returns collateral", async function () {
+      await advanceTimeForNextBlock(request.expiry / 2)
```
It's not clear to me why we need to advance the time

emizzle replied:
Makes sense. Maybe add a comment?

emizzle commented on test/Marketplace.test.js:
```diff
@@ -623,13 +644,37 @@ describe("Marketplace", function () {
+      const startBalance = await token.balanceOf(client.address)
+      await marketplace.withdrawFunds(slot.request)
+      const endBalance = await token.balanceOf(client.address)
+      expect(endBalance - startBalance).to.lt(
```
```suggestion
      expect(endBalance - startBalance).to.be.lt(
```

emizzle commented on test/Marketplace.test.js:
```diff
@@ -623,13 +644,37 @@ describe("Marketplace", function () {
+      )
+    })
+
+    it("withdraws to the client when request is cancelled", async function () {
```
Should there also be a test checking that cancelled requests pay out for slots that were filled before cancellation?

emizzle reviewed (approved):
Nice work! Overall a few suggestions and questions, but nothing blocking, so will approve now.

0x-r4bbit commented:
I'd be great if we could land #162 first, so we can sure these changes pass the rules (or maybe rules need to be changed).

AuHau commented:
> I'd be great if we could land #162 first, so we can sure these changes pass the rules (or maybe rules need to be changed).

For sure! I am waiting for Mark's review anyhow ;-)

markspanbroek commented on test/Marketplace.test.js:
```diff
@@ -778,17 +842,48 @@ describe("Marketplace", function () {
       )
     })
+
+    it.skip("when slot is freed and not filled, client will get refunded the freed slot's funds", async function () {
```
🔴 It's not entirely clear to me what requirement you're trying to test here. I'd expect the requirement to be that the client is given the funds that remain after the hosts are paid. So if you know the payout to the hosts, then the rest is for the client. I'd then also structure the test in that way:
1. fill a slot halfway through the expiry window
2. when contract is finished, payout the hosts
3. withdraw funds, and check that it equals the paid amount minus the payouts to the hosts

AuHau replied:
It needs rework agree, but I am trying to be here a bit more explicit in my assertions here, because doing assertions on "something is something minus something" is IMO quite week, because if there is bug in the calculation of the reward, then this would pass and it would not catch the sizing of the propotions...

Also, I am trying to test case where request does finish with not all slots filled and that the reward for that slot is returned.

markspanbroek replied:
Alright, I'll look at it again after you rework the test.

AuHau replied:
Done, please check again.

markspanbroek commented on test/Marketplace.test.js:
```diff
@@ -490,52 +490,53 @@ describe("Marketplace", function () {
 
-    it("pays to host reward address when contract has finished and returns collateral to host collateral address", async function () {
-      await waitUntilStarted(marketplace, request, proof, token)
+      await waitUntilStarted(marketplace, request, proof, token, [slot.index])
```
🟠 At a cursory glance, I'd expect this statement to wait until the slot with `slot.index` is started. Instead it does the opposite. You can solve this by making it a named parameter:

```suggestion
      await waitUntilStarted(marketplace, request, proof, token, { ignoreSlotIndexes: [slot.index] })
```

markspanbroek reviewed (commented):
Thanks for taking a stab at this @AuHau. I left a few comments.

0x-r4bbit commented:
Prover seems to be unhappy here as well btw. (yes the verify task still reports success, we need to get this fixed)

AuHau commented on contracts/Marketplace.sol:
```diff
@@ -373,24 +376,43 @@ contract Marketplace is Proofs, StateRetrieval, Endian {
+        _payoutAmount(requestId, requestExpiry(requestId));
+    } else if (state == RequestState.Failed) {
+      // For Failed requests the client is refunded whole amount.
+      context.fundsToReturnToClient = request.maxPrice();
```
Yeah makes sense. Although I feel like it is a fairly weak mitigation. This mainly could open attacks on the hosts before ending of the request to wiggle out of payment for them... I guess since we are gonna have limited duration for Requests this should be OK as the cost of the attack would be most probably higher than what you would wiggled out from. But maybe the request's duration limit should stay on the lower end "for ever"?

markspanbroek replied:
Good point. It's not a very practical attack anyway.

I would not be in favor of keeping the duration limit low for ever. It limits the usefulness of the network.

AuHau replied:
> I would not be in favor of keeping the duration limit low for ever. It limits the usefulness of the network.

I disagree about the limiting of the usefulness of the network. IMHO we can nicely plug in account abstraction for "recurring request renewal" to host requests "for ever". If we allow long requests in the system, then we will open doors to this attack that you mentioned before.

emizzle replied:
> If we allow long requests in the system, then we will open doors to this attack that you mentioned before.

30 days is long enough for this attack. A DDOS attack would only need to be long enough that it can force slots to be freed right near the end of the contract. The needed duration of the attack would depend on the proof frequency, however 30 days should be way more time than is needed.

I really like the idea of burning a percentage of the funds that would be returned to the client, however:
1. I see the cost of this attack as the mitigation needed, as there is no financial gain to be had from performing the attack. The only gain is disruption of the network. 
2. From an honest user's perspective, getting all funds returned in case the network fails to live up to its promises is a minimal level of consolation needed. If the user were to lose funds when the network fails (burned), I'm not sure how eager they'd be to come back and use Codex again.

AuHau replied:
> 30 days is long enough for this attack. A DDOS attack would only need to be long enough that it can force slots to be freed right near the end of the contract.

I am talking here about the client's attack, which could "wiggle out" of the payment for the stored data if he DDOSed the providers at the end of the storage requests and hence got returned the full reward amount.

What I am mentioning here is that if the limit of storage request duration is low, then the price for the request is low, which I would argue that the gain from such an attack should be small enough to justify a DDOS attack on several node providers. 

ChatGPT (hence take it with a grain of salt) says the cheapest DDOS starts at 10$ per hour. Hence, you would need at least 24 hours for several days in order for the providers to get kicked out of the request and make it fail. So the cost to kick out a single provider starts at least on: `$10 * 24 * 4 = $960` (the 4 is the arbitrary number for how many slashes it takes before the node is kicked out). Based on the tolerance setting, there might need to be more providers kicked out as well.

emizzle replied:
I'm pretty sure we're saying the same thing 😅  I was only trying to point out that our "low duration" of 30 days is already long enough to allow for DDOS attacks and so this is somewhat of a moot point in arguing for keeping contract duration low.

> the gain from such an attack
Technically, there is no gain except "free" storage and network disruption, as there is no flow of *additional* funds to the client. They would simply get their own funds returned to them. Well they might get reimbursed for the gas cost associated with creating the request, but there's no way that would exceed the cost of a DDOS attack. It'll almost definitely be cheaper to actually pay for the storage 😄

AuHau commented on contracts/Marketplace.sol:
```diff
@@ -373,24 +376,43 @@ contract Marketplace is Proofs, StateRetrieval, Endian {
+      // When requests are cancelled, funds earmarked for payment for the duration
+      // between request expiry and request end (for every slot that was filled), should be returned to the client.
+      // Update `fundsToReturnToClient` to reflect this.
+      context.fundsToReturnToClient +=
```
> Doesn't this mean that the amount of funds that a client receives depends on when it calls withdrawFunds().

No, it does not depend on that. You are most probably referencing the `context.slotsFilled` right? That variable is not lowered when calling `freeSlot` in states **Failed/Cancelled/Finished**, only when the node would call that function during "running" of the Request, which again would not effect this function as it can be called only in those states...

markspanbroek replied:
No, I didn't mean `context.slotsFilled`. I was thinking about the additions to `context.fundsToReturnToClient` in `freeSlot`. But looking through the code again, I realize that this only happens in `_forciblyFreeSlot()`, and that can only happen while the slot is Filled.

AuHau commented on contracts/Marketplace.sol:
```diff
@@ -373,24 +376,43 @@ contract Marketplace is Proofs, StateRetrieval, Endian {
+        state == RequestState.Finished,
+      "Invalid state"
+    );
+    require(context.fundsToReturnToClient != 0, "Nothing to withdraw");
```
But a good idea is that I should have a test that will verify the double-spend protection...

markspanbroek replied:
You could consider adding a boolean `context.clientPaid` to make it more explicit.

AuHau replied:
IMHO this mechanism is sufficient if properly documented, but at the same time I don't have strong opinions about it either, so if you think I should go with the flag approach, then let me know.

emizzle replied:
I agree that a flag would be more explicit, albeit it does incur a higher gas cost for storing additional state.

AuHau commented on contracts/Marketplace.sol:
```diff
@@ -373,24 +376,43 @@ contract Marketplace is Proofs, StateRetrieval, Endian {
     assert(_token.transfer(withdrawRecipient, amount));
+
+    // We zero out the funds tracking in order to prevent double-spends
+    context.fundsToReturnToClient = 0;
```
I don't think it is necessary. Also the reason why I would avoid this approach is because we will lose information about how the request terminated if we move from the current terminating states into "one that catches them all".

markspanbroek replied:
Good  point, we still need the distinction on how a request ended to handle `freeSlot()` correctly.

AuHau commented on test/Marketplace.test.js:
```diff
@@ -778,17 +842,48 @@ describe("Marketplace", function () {
       )
     })
+
+    it.skip("when slot is freed and not filled, client will get refunded the freed slot's funds", async function () {
```
Haha yeah, because it does not work 🤪 I wanted to return to this test to rewrite it properly.

AuHau commented:
I have implemented your feedback and reworked the marketplace tooling in tests where `waitUntilStarted()` now returns the array of expected payouts for each filled slot and I use that to make the calculations - IMHO this looks much better and I would say is good to go.

markspanbroek commented on test/Marketplace.test.js:
```diff
@@ -739,7 +785,36 @@ describe("Marketplace", function () {
+      // then the upper bound to how much the client gets returned is the cumulative reward for all the
+      // slots for expiry window. This limit is "inclusive" because it is possible that all slots are filled
+      // at the time of expiry and hence the user would get the full "expiry window" reward back.
+      expect(endBalancePayout - startBalancePayout).to.be.lte(
```
Perhaps also a check here that the payout is > 0?

markspanbroek commented on test/marketplace.js:
```diff
@@ -14,12 +14,38 @@ async function waitUntilCancelled(request) {
+  return payouts
+}
+
+function slotsArrayExcept(request, ignoredIndex) {
```
This is not used anywhere, as far as I can tell.

markspanbroek reviewed (approved):
Thanks @AuHau! The tests are very nice now.
